README.md
[![Build Status](https://travis-ci.org/stephanepericat/express-db.png?branch=master)](https://travis-ci.org/stephanepericat/express-db)

#express-db 0.0.2#

A lightweight local key-value store that integrates with your express.js application.

**WARNING** This module is in very early stage of development. Serious issues, like performance, scalability or security may occur. Use at your **OWN RISK**!
If you want to help improve express-db, feel free to contribute by forking and sending pull requests!

##Installation##

[![NPM](https://nodei.co/npm/express-db.png)](https://nodei.co/npm/express-db/)

##Usage##

###Create you own db file###

The first thing to do is to create your own db file for backup. It will need to be a json file, with the following default structure:

	{
	  "info": {
	    "version": "0.0.2",
	    "name": "demo DB"
	  },
	  "data": {

	  }
	}

'version' should be set to the version of express-db you are using, and 'name' should be the name of you db.

You will then need to define it in the express-db constructor:

	xdb.init("my custom DB", {
		file: './my.custom.db.json'
	});

###Integrate with express.js 3.x###

	var express = require('express'),
		app = express(),
		xdb = require('express-db'),
		server = require('http')
			.createServer(app)
			.listen(8081);

	app.use(xdb.init("demo DB", {
		file: './db/demo.db.json', //your db file
		restrictAccess: false, //restrict access via browser
		autoSave: true, //autosave enabled
		backupInterval: 60000, //interval in ms,
		viewCaching: true //cache views then requested
	}));

###API Reference###

####init(dbname, options)####

the init method, to be passed to express for hook.

*Options*:

 + file: (*required*) path to your own db file
 + restrictAccess: (*optional*) true or false, restricts the access to the db on the client side
 + autoSave: (*optional*) enables automatic backup of the databse
 + backupInterval: (*optional*) the frequency (in ms) of backups
 + viewCaching: (*optional*) true or false, caches the view result on browser call

*Example*:

	xdb.init("demo DB", {
		file: './db/demo.db.json', //your db
		restrictAccess: false, //restrict access via browser
		autoSave: true, //autosave enabled
		backupInterval: 60000, //interval in ms,
		viewCaching: true //cache views then requested
	});

####info()###

Returns info on the current database.

	xdb.info() // => {"version" : "0.0.2", "name": "demo DB"}

####get(key)####

Gets the value for a given key.

	xdb.get('foo'); // => 'bar'

####set(key, value)####

Sets a value for a give key.

	xdb.set('foo', 'bar');

####remove(key)####

Removes a given key.

	xdb.remove('foo');

####backup(callback)####

Forces an immediate backup to file of the current database.

	xdb.backup(function(err) {
		if(err) throw err;
		else console.log('DB backed up at %s', new Date().toString());
	});

####search(query)####

Search through the JSON structure. Read the [json-query documentation](https://github.com/mmckegg/json-query/blob/master/README.md) for more details on query possibilities.

	xdb.search('stats.comments_lookup[{stats.page.id}]', function(res) {
		console.log("search result: ", res);
	});

####createView(name, query)####

Creates a view exposbale to the client side of the express.js application. Query system is also based on [json-query](https://github.com/mmckegg/json-query/blob/master/README.md).

	xdb.createView('comments', 'stats.comments_lookup[{stats.page.id}]');

####getViews()####

Returns the views defined for the current database.

	xdb.getViews(); // => {'comments': {'query': 'stats.comments_lookup[{stats.page.id}]', cache: null}, ...}

##Test##

	npm test

##Performance##

Tested on MacBook Air (1.7 Ghz Intel Core i5, 4Gb 1600Mhz DDR3):

	STARTED SET TEST: 1373816191605
	FINISHED SET TEST: 1373816192791
	100,000 records set in : 1186 milliseconds
	slept for 2 seconds
	STARTED GET TEST: 1373816194793
	FINISHED GET TEST: 1373816195584
	100,000 records get in : 791 milliseconds

You can run the test on your own computer by typing: `node demo/test.js`

##License##

The MIT License (MIT)

Copyright (c) 2013 Stephane P. Pericat

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.


CHANGELOG.md
#Changelog#

 + 0.0.2: various bugfixes, update docs
 + 0.0.1: first version

demo/test.js
var express = require('express'),
	app = express(),
	xdb = require('../lib/express-db');

app.use(xdb.init("stress test DB", {
	file: './bin/test.db.json',
	autoSave: true
}));

var beginSet = new Date();
console.log("STARTED SET TEST: %s", beginSet.getTime().toString());
for(var i = 0; i < 100000; i++) {
	xdb.set('test:'+i, i);
}
var endSet = new Date();
console.log("FINISHED SET TEST: %s", endSet.getTime().toString());
console.log("100,000 records set in : %s milliseconds", (endSet.getTime() - beginSet.getTime()).toString());

setTimeout(function() {
	console.log("slept for 2 seconds");
	var beginGet = new Date();
	console.log("STARTED GET TEST: %s", beginGet.getTime().toString());
	for(var i = 0; i < 100000; i++) {
		xdb.get('test:'+i, i);
	}
	var endGet = new Date();
	console.log("FINISHED GET TEST: %s", endGet.getTime().toString());
	console.log("100,000 records get in : %s milliseconds", (endGet.getTime() - beginGet.getTime()).toString());
}, 2000);

lib/expressDB.js
var nconf = require('nconf'),
	colors = require('colors'),
	dbStorage = nconf.use('file', {file: './bin/db.json'}),
	instanceName,
	autoSave = false,
	backupInterval = 300000;

//load the db
dbStorage.load();

//private methods
function enableAutosave(backupInterval) {
	setInterval(function() {
		dbStorage.save(function(err) {
			if(err) {
				console.log("<ERROR> %s : expressDB save error : %s", new Date().toString(), err.toString());
			} else {
				console.log("<LOG> %s : expressDB : %s saved successfully.", new Date().toString(), instanceName);
			}
		});
	}, backupInterval);
};

//module
module.exports = {
	init: function(iName, options) {
		if(options && options.file) {
			dbStorage = nconf.use('file', {file: options.file});
			dbStorage.load();
		}
		instanceName = (iName) ? iName : "expressDB";
		autoSave = (options && options.autoSave) ? options.autoSave : true;
		if(autoSave) {
			backupInterval = (options && options.backupInterval) ? options.backupInterval : 300000; //default is every 5 minutes
			enableAutosave(backupInterval);
		}

		if(!dbStorage.get('info:name') || (instanceName !== dbStorage.get('info:name'))) {
			dbStorage.set('info:name', instanceName);
			dbStorage.save(function() {
				console.log("<LOG> %s : expressDB db name set to: %s", new Date().toString(), instanceName);
			});
		}
		
		return function(req, res, next) {
			if(/db/.test(req.url)) {
				if('/db' === req.url) {
					res.writeHead(200, {'Content-Type': 'application/json'});
					res.end(JSON.stringify({name:instanceName, version: "0.0.1", status: "OK"}));
				} else {
					//close external db access
					res.writeHead(200, {'Content-Type': 'application/json'});
					res.end(JSON.stringify("forbidden"));
				}

				console.log('<LOG> %s : expressDB query : %s', new Date().toString(), req.url);
			} else {
				next();
			}
		};
	},
	info: function() {
		return (dbStorage.get('info')) ? dbStorage.get('info') : {};
	},
	get: function(key) {
		var value = dbStorage.get('data:' + key);
		return (value) ? value : false;
	},
	set: function(key, value) {
		dbStorage.set('data:' + key, value);
		return true;
	},
	backup: function(callback) {
		dbStorage.save(callback);
	}
};

lib/express-db.js
var nconf = require('nconf'),
	colors = require('colors'),
	jsonQuery = require('json-query'),
	dbStorage = nconf.use('file', {file: './bin/db.json'}),
	instanceName,
	searchContext,
	views = {},
	viewCache = true,
	restrictedAccess = false,
	autoSave = false,
	backupInterval = 300000;

//load the db
dbStorage.load();

//private methods
function enableAutosave(backupInterval) {
	setInterval(function() {
		dbStorage.save(function(err) {
			if(err) {
				sendLog('error', 'express-db save error : %s', err.toString());
			} else {
				sendLog('log', 'express-db : %s saved successfully', instanceName.green);
			}
		});
	}, backupInterval);
};

function sendLog(msgType, msg /*, arg1, arg 2,... arg N*/) {
	var type = msgType,
		str,
		args = [];

	switch(type) {
		case "log":
			str = '[LOG] :: '.blue + new Date().toString().grey + ' :: express-db (' + instanceName.green + ') :: '.grey + msg.grey; 
		break;
		case "error":
			str = '[ERROR] :: '.red.bold + new Date().toString().grey + ' :: express-db (' + instanceName.green + ') :: '.grey + msg.red; 
		break;
		default:
			str = '[LOG] :: '.blue + new Date().toString().grey + ' :: express-db (' + instanceName.green + ') :: '.grey + msg.grey; 
		break;
	}

	args.push(str);

	if(arguments.length > 2) {
		for(var i = 2; i < arguments.length; i++) {
			args.push(arguments[i]);
		}
	}
	console.log.apply(this, args);
}

//module
module.exports = {
	init: function(iName, options) {
		if(options && options.file) {
			dbStorage = nconf.use('file', {file: options.file});
			dbStorage.load();
		}

		instanceName = (iName) ? iName : "expressDB";
		searchContext = dbStorage.get('data');
		restrictedAccess = (options && options.restrictAccess) ? options.restrictAccess : false;
		viewCache = (options && options.viewCaching) ? options.viewCaching : true;

		autoSave = (options && options.autoSave) ? options.autoSave : true;
		if(autoSave) {
			backupInterval = (options && options.backupInterval) ? options.backupInterval : 300000; //default is every 5 minutes
			enableAutosave(backupInterval);
		}

		if(!dbStorage.get('info:name') || (instanceName !== dbStorage.get('info:name'))) {
			dbStorage.set('info:name', instanceName);
			dbStorage.save(function() {
				sendLog('log', 'expressDB db name set to: %s', instanceName);
			});
		}
		
		return function(req, res, next) {
			if(/db/.test(req.url)) {
				if('/db' === req.url) {
					res.writeHead(200, {'Content-Type': 'application/json'});
					res.end(JSON.stringify({name:instanceName, version: "0.0.1", status: "OK"}));
				} else {
					if(restrictedAccess) {
						//close external db access
						res.writeHead(403, {'Content-Type': 'application/json'});
						res.end(JSON.stringify({"express-db": "access forbidden"}));
					} else {
						if(/\/db\/view\//.test(req.url)) {
							var viewName = /\/db\/view\/(.*)/.exec(req.url);
							if(viewName && viewName[1] && views[viewName[1]]) {
								var viewQuery = jsonQuery(views[viewName[1]].query, {rootContext: searchContext}),
									viewData = (viewQuery && viewQuery.value) ? viewQuery.value : {};

								if(viewCache) {
									views[viewName[1]].cache = (viewQuery && viewQuery.value) ? viewQuery.value : null;
								}

								res.writeHead(200, {'Content-Type': 'application/json'});
								res.end(JSON.stringify({"view": {"name": viewName[1], "data": viewData}}));
							} else {
								res.writeHead(404, {'Content-Type': 'application/json'});
								res.end(JSON.stringify({"express-db": "not found"}));
							}
						} else {
							res.writeHead(404, {'Content-Type': 'application/json'});
							res.end(JSON.stringify({"express-db": "not found"}));
						}
					}
				}

				sendLog('log', 'express-db query : %s', req.url.yellow);
			} else {
				next();
			}
		};
	},

	info: function() {
		return (dbStorage.get('info')) ? dbStorage.get('info') : {};
	},

	get: function(key) {
		var value = dbStorage.get('data:' + key);
		return (value) ? value : false;
	},

	set: function(key, value) {
		dbStorage.set('data:' + key, value);
		return true;
	},

	remove: function(key, callback) {
		dbStorage.clear('data:' + key, callback);
	},

	backup: function(callback) {
		dbStorage.save(callback);
	},

	search: function(query, callback) {
		var searchResult = jsonQuery(query, {rootContext: searchContext});
		callback.call(this, searchResult);
	},

	createView: function(viewName, query) {
		views[viewName] = {
			query: query,
			cache: null
		};
		return true;
	},

	getViews: function() {
		return views;
	}
};

demo/app.js
var express = require('express'),
	app = express(),
	xdb = require('../lib/express-db'),
	server = require('http')
		.createServer(app)
		.listen(8081);

app.use(xdb.init("demo DB", {
	file: './bin/demo.db.json', //custom db file, if necessary
	restrictAccess: false, //restrict access via browser
	autoSave: true, //autosave enabled
	backupInterval: 60000, //interval in ms,
	viewCaching: true //cache views then requested
}));

//set
xdb.set('foo', 'bar');
//get
console.log('foo ? %s', xdb.get('foo'));
//remove
xdb.remove('foo', function(err) {
	console.log('is foo still here ? %s', xdb.get('foo'));
});
//search for the comments for page id 'page_1'
xdb.search('stats.comments_lookup[{stats.page.id}]', function(res) {
	console.log("search result: ", res);
});
//create view
xdb.createView('blogposts', 'posts');
xdb.createView('comments', 'stats.comments_lookup[{stats.page.id}]');
//get views
console.log(xdb.getViews());

// express stuff
app.get('/', function(req, res) {
	res.end("hello world!");
});